File: Sources/ViewModels/GroupListViewModel.swift
import Foundation
import WhistleCore
import Combine
import MDKBindings

/// Drives the Chat tab group list — observes `MarmotService.groups`.
@MainActor
final class GroupListViewModel: ObservableObject {

    // MARK: - Published state

    @Published private(set) var groups: [GroupListItem] = []
    @Published var showCreateGroup = false
    @Published var showJoinGroup = false
    /// Pre-populated invite code delivered via deep link / QR scan / NFC.
    @Published var pendingJoinCode: String?

    // MARK: - Dependencies

    private let marmot: MarmotService
    private let mls: MLSService
    private let displayName: () -> String
    let pendingInviteStore: PendingInviteStore
    let pendingLeaveStore: PendingLeaveStore
    let healthTracker: GroupHealthTracker
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Unread tracking

    private static let lastReadKey = "groupLastReadTimestamps"
    private static let lastChatKey  = "groupLastChatTimestamps"

    /// Per-group epoch timestamp of the last time the user viewed the chat.
    private var lastReadTimestamps: [String: TimeInterval] {
        get { UserDefaults.standard.dictionary(forKey: Self.lastReadKey) as? [String: TimeInterval] ?? [:] }
        set { UserDefaults.standard.set(newValue, forKey: Self.lastReadKey) }
    }

    /// Per-group epoch timestamp of the last chat message received (chat-only — not location/nickname/etc.).
    /// Used by refreshItems so that non-chat MLS events don't re-trigger the unread indicator.
    private var lastChatTimestamps: [String: TimeInterval] {
        get { UserDefaults.standard.dictionary(forKey: Self.lastChatKey) as? [String: TimeInterval] ?? [:] }
        set { UserDefaults.standard.set(newValue, forKey: Self.lastChatKey) }
    }

    /// Call when the user opens a group chat to clear the unread indicator.
    func markAsRead(groupId: String) {
        lastReadTimestamps[groupId] = Date().timeIntervalSince1970
        // Update the list to clear bold state
        if let idx = groups.firstIndex(where: { $0.id == groupId }), groups[idx].hasUnread {
            groups[idx].hasUnread = false
        }
    }

    // MARK: - Item model

    struct GroupListItem: Identifiable, Hashable {
        let id: String          // mlsGroupId
        let name: String
        let memberCount: Int
        let lastActivity: Date?
        let isActive: Bool
        var hasUnread: Bool = false
    }

    // MARK: - Init

    init(
        marmot: MarmotService,
        mls: MLSService,
        pendingInviteStore: PendingInviteStore,
        pendingLeaveStore: PendingLeaveStore,
        displayName: @escaping () -> String = { "" }
    ) {
        self.marmot = marmot
        self.mls = mls
        self.pendingInviteStore = pendingInviteStore
        self.pendingLeaveStore = pendingLeaveStore
        self.healthTracker = marmot.healthTracker
        self.displayName = displayName

        marmot.$groups
            .debounce(for: .milliseconds(200), scheduler: DispatchQueue.main)
            .sink { [weak self] groups in
                Task { await self?.refreshItems(from: groups) }
            }
            .store(in: &cancellables)

        // When a new chat message arrives, mark that group as unread immediately
        // and persist the timestamp so refreshItems can use it instead of MDK's
        // lastMessageAt (which advances for location/nickname events too).
        marmot.$lastChatMessageGroupId
            .compactMap { $0 }
            .sink { [weak self] groupId in
                guard let self else { return }
                var stamps = self.lastChatTimestamps
                stamps[groupId] = Date().timeIntervalSince1970
                self.lastChatTimestamps = stamps
                if let idx = self.groups.firstIndex(where: { $0.id == groupId }) {
                    self.groups[idx].hasUnread = true
                }
            }
            .store(in: &cancellables)

        // Merge child objectWillChange and debounce to avoid cascading renders.
        Publishers.Merge3(
            pendingInviteStore.objectWillChange.map { _ in () }.eraseToAnyPublisher(),
            pendingLeaveStore.objectWillChange.map { _ in () }.eraseToAnyPublisher(),
            healthTracker.objectWillChange.map { _ in () }.eraseToAnyPublisher()
        )
        .debounce(for: .milliseconds(50), scheduler: DispatchQueue.main)
        .sink { [weak self] _ in self?.objectWillChange.send() }
        .store(in: &cancellables)
    }

    // MARK: - Refresh

    func refresh() async {
        await marmot.refreshGroups()
    }

    private func refreshItems(from mdkGroups: [Group]) async {
        // Fetch member counts first — this is the only async work.
        var memberCounts: [String: Int] = [:]
        for group in mdkGroups {
            memberCounts[group.mlsGroupId] = (try? await mls.getMembers(groupId: group.mlsGroupId).count) ?? 0
        }
        // Read timestamps AFTER all awaits so any markAsRead calls that happened
        // during suspension are reflected — avoids showing already-read groups as unread.
        // Use lastChatTimestamps (chat-only) rather than group.lastMessageAt, which MDK
        // advances for every MLS event including location and nickname updates.
        let readTimestamps = lastReadTimestamps
        let chatTimestamps = lastChatTimestamps
        var items: [GroupListItem] = []
        for group in mdkGroups {
            let lastMessageEpoch = group.lastMessageAt.map { TimeInterval($0) }
            let lastChatEpoch = chatTimestamps[group.mlsGroupId]
            let lastRead = readTimestamps[group.mlsGroupId] ?? 0
            let hasUnread = lastChatEpoch.map { $0 > lastRead } ?? false
            items.append(GroupListItem(
                id: group.mlsGroupId,
                name: group.name.isEmpty ? "Unnamed Group" : group.name,
                memberCount: memberCounts[group.mlsGroupId] ?? 0,
                lastActivity: lastMessageEpoch.map { Date(timeIntervalSince1970: $0) },
                isActive: group.isActive,
                hasUnread: hasUnread
            ))
        }
        self.groups = items.filter { !pendingLeaveStore.contains($0.id) }

        // Clean up pending leaves for groups that no longer exist
        // (admin processed the removal).
        let activeIds = Set(items.map(\.id))
        pendingLeaveStore.removeResolved(activeGroupIds: activeIds)
    }

    // MARK: - Actions

    func createGroup(name: String) async throws -> String {
        let relays = marmot.activeRelayURLs
        let groupId = try await marmot.createGroup(name: name, relays: relays)

        // Broadcast our display name so other members see it immediately
        let dn = displayName()
        if !dn.isEmpty {
            try? await marmot.sendNicknameUpdate(name: dn, toGroup: groupId)
        }

        return groupId
    }

    /// Poll relays for gift-wrap events that may have been missed by the
    /// real-time subscription. Called by JoinGroupView while waiting for Welcome.
    func fetchMissedWelcomes() async {
        await marmot.fetchMissedGiftWraps()
    }

    /// Send a leave request to the group and mark it as "Leaving…" locally.
    func requestLeaveGroup(id: String) async {
        do {
            try await marmot.sendLeaveRequest(groupId: id)
            pendingLeaveStore.add(id)
        } catch {
            FMFLogger.chat.error("Failed to request leave for group \(id): \(error)")
        }
    }

    /// Force relay reconnection after MPC / NearbyShare activity.
    func forceReconnectRelays() async {
        await marmot.forceReconnectRelays()
    }

    func joinGroup(inviteCode: String) async throws {
        // Decode first so we can extract the group hint for pending state.
        let invite = try InviteCode.decode(from: inviteCode)

        try await marmot.acceptInvite(inviteCode)

        // If the user previously left this group, clear the stale pending leave
        // marker so the group reappears once Welcome is accepted.
        pendingLeaveStore.remove(invite.groupId)

        // Record as pending — will be auto-removed when Welcome arrives.
        pendingInviteStore.add(PendingInvite(
            groupHint: invite.groupId,
            inviterNpub: invite.inviterNpub
        ))
    }
}

File: android/app/src/main/java/org/findmyfam/viewmodels/GroupListViewModel.kt
package org.findmyfam.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import build.marmot.mdk.Group
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.findmyfam.models.AppSettings
import org.findmyfam.shared.models.InviteCode
import org.findmyfam.shared.models.PendingInvite
import org.findmyfam.services.*
import timber.log.Timber
import javax.inject.Inject

/**
 * Drives the group list screen -- observes MarmotService groups,
 * exposes GroupListItem list, create/join/leave actions.
 */
@HiltViewModel
class GroupListViewModel @Inject constructor(
    private val marmotService: MarmotService,
    private val mlsService: MLSService,
    private val settings: AppSettings,
    private val pendingInviteStore: PendingInviteStore,
    private val pendingLeaveStore: PendingLeaveStore,
    private val pendingWelcomeStore: PendingWelcomeStore
) : ViewModel() {

    // --- Item model ---

    data class GroupListItem(
        val id: String,          // mlsGroupId
        val name: String,
        val memberCount: Int,
        val lastActivity: Long?, // epoch seconds
        val isActive: Boolean,
        val hasUnread: Boolean = false
    )

    // --- Published state ---

    private val _groups = MutableStateFlow<List<GroupListItem>>(emptyList())
    val groups: StateFlow<List<GroupListItem>> = _groups.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    val pendingInvites: StateFlow<List<PendingInvite>> = pendingInviteStore.pendingInvites
    val pendingLeaves: StateFlow<Set<String>> = pendingLeaveStore.pendingLeaves
    val pendingWelcomes: StateFlow<List<PendingWelcomeItem>> = pendingWelcomeStore.pendingWelcomes

    private val _pendingAdminActionGroupIds = MutableStateFlow<Set<String>>(emptySet())
    val pendingAdminActionGroupIds: StateFlow<Set<String>> = _pendingAdminActionGroupIds.asStateFlow()

    /** Dismiss a stale pending invite that will never be accepted. */
    fun cancelPendingInvite(groupHint: String) {
        pendingInviteStore.remove(groupHint)
    }

    /** Accept an unsolicited welcome after user approval. */
    fun approvePendingWelcome(mlsGroupId: String) {
        viewModelScope.launch {
            try {
                marmotService.approvePendingWelcome(mlsGroupId)
            } catch (e: Exception) {
                _error.value = e.message
                Timber.e("Failed to approve welcome for group $mlsGroupId: $e")
            }
        }
    }

    /** Decline an unsolicited welcome. */
    fun declinePendingWelcome(mlsGroupId: String) {
        viewModelScope.launch {
            try {
                marmotService.declinePendingWelcome(mlsGroupId)
            } catch (e: Exception) {
                _error.value = e.message
                Timber.e("Failed to decline welcome for group $mlsGroupId: $e")
            }
        }
    }

    val unhealthyGroupIds: StateFlow<Set<String>> = marmotService.healthTracker.unhealthyGroupIds

    init {
        // Observe MarmotService groups and rebuild list items
        viewModelScope.launch {
            marmotService.groups.collect { mdkGroups ->
                refreshItems(mdkGroups)
            }
        }
        // A pendingWelcomes emission alone doesn't recompute _groups -- its
        // pendingWelcomeIds filter only runs inside refreshItems. Without this, a
        // group added to pendingWelcomeStore (e.g. by a resync re-add) keeps
        // showing its stale prior entry -- inactive row AND a new "Accept" row for
        // the same group -- until an unrelated marmotService.groups emission
        // happens to rerun the filter.
        viewModelScope.launch {
            pendingWelcomeStore.pendingWelcomes.collect {
                refreshItems(marmotService.groups.value)
            }
        }
        // When a new chat message arrives, persist the timestamp and mark unread immediately.
        // Persisting here means refreshItems can use a chat-only timestamp rather than
        // MDK's lastMessageAt, which advances for location/nickname events too.
        viewModelScope.launch {
            marmotService.lastChatMessageGroupId.collect { change ->
                val groupId = change?.first
                if (groupId != null) {
                    settings.recordChatMessage(groupId)
                    _groups.value = _groups.value.map {
                        if (it.id == groupId) it.copy(hasUnread = true) else it
                    }
                }
            }
        }
    }

    /** Mark a group as read — call when the user opens a group chat. */
    fun markAsRead(groupId: String) {
        settings.markGroupAsRead(groupId)
        // Update list immediately to clear bold
        _groups.value = _groups.value.map {
            if (it.id == groupId) it.copy(hasUnread = false) else it
        }
    }

    // --- Refresh ---

    fun refresh() {
        viewModelScope.launch {
            _isRefreshing.value = true
            try {
                marmotService.refreshGroups()
            } finally {
                _isRefreshing.value = false
            }
        }
    }

    private suspend fun refreshItems(mdkGroups: List<Group>) {
        // Fetch member counts first — this is the only async work.
        val memberCounts = mutableMapOf<String, Int>()
        for (group in mdkGroups) {
            memberCounts[group.mlsGroupId] = try {
                mlsService.getMembers(group.mlsGroupId).size
            } catch (_: Exception) { 0 }
        }
        // Read timestamps AFTER all awaits so any markGroupAsRead calls that happened
        // during suspension are reflected — avoids showing already-read groups as unread.
        val items = mdkGroups.map { group ->
            val lastMessageEpoch = group.lastMessageAt?.toLong()
            val lastChatEpoch = settings.getLastChatTimestamp(group.mlsGroupId)
            val lastRead = settings.getLastRead(group.mlsGroupId)
            val hasUnread = lastChatEpoch != null && lastChatEpoch > lastRead
            GroupListItem(
                id = group.mlsGroupId,
                name = group.name.ifEmpty { "Unnamed Group" },
                memberCount = memberCounts[group.mlsGroupId] ?: 0,
                lastActivity = lastMessageEpoch,
                isActive = group.state == "active",
                hasUnread = hasUnread
            )
        }.filter { !pendingLeaveStore.contains(it.id) }

        val pendingWelcomeIds = pendingWelcomeStore.pendingWelcomes.value.map { it.mlsGroupId }.toSet()

        _groups.value = items.filter { it.id !in pendingWelcomeIds }

        // Recompute admin action badges
        _pendingAdminActionGroupIds.value = settings.pendingLeaveRequests
            .filter { it.value.isNotEmpty() }.keys

        // Clean up pending leaves for groups that no longer exist
        val activeIds = mdkGroups.map { it.mlsGroupId }.toSet()
        pendingLeaveStore.removeResolved(activeIds)
    }

    // --- Actions ---

    fun createGroup(name: String, onCreated: (String) -> Unit = {}) {
        viewModelScope.launch {
            try {
                _error.value = null
                val relays = marmotService.activeRelayUrls
                val groupId = marmotService.createGroup(name = name, relays = relays)

                // Broadcast our display name so other members see it immediately
                val dn = settings.displayName
                if (dn.isNotEmpty()) {
                    try {
                        marmotService.sendNicknameUpdate(name = dn, groupId = groupId)
                    } catch (_: Exception) { }
                }

                onCreated(groupId)
            } catch (e: Exception) {
                _error.value = e.message
                Timber.e("Failed to create group: $e")
            }
        }
    }

    fun joinGroup(inviteCode: String) {
        viewModelScope.launch {
            try {
                _error.value = null
                val invite = InviteCode.fromUri(inviteCode)
                val rawCode = invite.encode()

                marmotService.acceptInvite(rawCode)

                // If the user previously left this group, clear the stale pending leave
                pendingLeaveStore.remove(invite.groupId)

                // Record as pending -- will be auto-removed when Welcome arrives
                pendingInviteStore.add(
                    PendingInvite(
                        groupHint = invite.groupId,
                        inviterNpub = invite.inviterNpub
                    )
                )

                // Poll for the Welcome gift-wrap — the admin may add us at any
                // point in the next 2 minutes. Mirrors iOS JoinGroupView polling.
                pollForWelcome(invite.groupId)
            } catch (e: Exception) {
                _error.value = e.message
                Timber.e("Failed to join group: $e")
            }
        }
    }

    /**
     * Poll fetchMissedGiftWraps every 2s for up to 2 minutes, checking
     * whether the expected group has appeared in the groups list.
     */
    private fun pollForWelcome(expectedGroupId: String) {
        viewModelScope.launch {
            for (i in 0 until 60) {
                kotlinx.coroutines.delay(2000)
                try {
                    marmotService.fetchMissedGiftWraps()
                } catch (_: Exception) { }
                // Check if the group appeared
                if (_groups.value.any { it.id == expectedGroupId }) {
                    Timber.i("Welcome received for group $expectedGroupId after ${(i + 1) * 2}s")
                    return@launch
                }
            }
            Timber.w("Welcome polling timed out for group $expectedGroupId")
        }
    }

    fun requestLeaveGroup(groupId: String) {
        viewModelScope.launch {
            try {
                marmotService.sendLeaveRequest(groupId = groupId)
                pendingLeaveStore.add(groupId)
            } catch (e: Exception) {
                Timber.e("Failed to request leave for group $groupId: $e")
            }
        }
    }

    fun fetchMissedWelcomes() {
        viewModelScope.launch {
            marmotService.fetchMissedGiftWraps()
        }
    }
}
